### PR TITLE
rate limit

### PR DESCRIPTION
**Pull request by ermalkaleci**

### CODE REVIEW

File: src/extensions/server/mod.rs [hunk 1]
```diff
@@ -50,12 +53,35 @@ pub struct ServerConfig {
...
+    pub rate_limit: Option<RateLimitConfig>,
 }
 
 fn default_request_timeout_seconds() -> u64 {
     120
 }
 
+#[derive(Deserialize, Default, Debug, Copy, Clone)]
+#[serde(rename_all = "snake_case")]
+pub enum Period {
```
[the remark this answers is not in the document]
Comment (ermalkaleci) [reply src/extensions/server/mod.rs:66]: Yes we can

File: src/extensions/server/mod.rs [hunk 2]
```diff
@@ -50,12 +53,35 @@ pub struct ServerConfig {
...
+    Minute,
+    Hour,
+}
+
+#[derive(Deserialize, Debug, Copy, Clone, Default)]
+pub struct RateLimitConfig {
+    pub burst: u32,
+    pub period: Period,
+    #[serde(default = "default_jitter_millis")]
+    pub jitter_millis: u64,
```
[the remark this answers is not in the document]
Comment (ermalkaleci) [reply src/extensions/server/mod.rs:78]: Random additional delay so delayed request won't be fired at the same time. If jitter 100ms they are spread between wait_time + rand(0-100)ms

File: config.yml [hunk 3]
```diff
@@ -23,6 +23,9 @@ extensions:
       - path: /liveness
         method: chain_getBlockHash
     cors: all
+    rate_limit:
+      burst: 20
```
[the remark this answers is not in the document]
Comment (ermalkaleci) [reply config.yml:27]: Yes sure. This is more an initial version

File: src/extensions/server/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (xlc) [reply src/extensions/server/mod.rs:66]: in that case make an input like `period_seconds` and we don't need this enum

File: src/extensions/server/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (xlc) [reply src/extensions/server/mod.rs:78]: I see. looks useful but need comments

File: config.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (xlc) [reply config.yml:27]: what I am asking is to make the config to be something like

```rs
rate_limit: Vec<RateLimitRule>

struct RateLimitRule {
  key: Vec<RateLimitKey>,
  period_seconds: u32,
  quota: u32,
}

enum RateLimitKey {
  Connection.
  IP,
  Unit,
}

`

File: config.yml [hunk 4]
```diff
@@ -23,6 +23,9 @@ extensions:
       - path: /liveness
         method: chain_getBlockHash
     cors: all
+  rate_limit: # 20 requests per second
```
Comment (xlc) [inline config.yml:26]:
```suggestion
  rate_limit: # 20 requests per second per connection
```

**Review by xlc — APPROVED** [review]

**Pull request merged**